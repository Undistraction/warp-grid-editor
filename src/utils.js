import { curry, reduce } from 'ramda'
import { isInteger } from 'ramda-adjunct'

import { CORNER_POINTS } from './const'
import { expandBoundingCurvesCornerControlPoints } from './utils/boundingCurves'
import { clampNumberBetween } from './utils/math'
import { getRandomRectangleBounds } from './utils/random'

// -----------------------------------------------------------------------------
// Const
// -----------------------------------------------------------------------------

const SHAPE_MIN_DISTANCE_FROM_EDGE = 100

// -----------------------------------------------------------------------------
// Utils
// -----------------------------------------------------------------------------

const getCornerPoints = (x, y, width, height) => {
  const rightBounds = x + width
  const bottomBounds = y + height

  return {
    topLeft: { x, y },
    topRight: { x: rightBounds, y },
    bottomRight: { x: rightBounds, y: bottomBounds },
    bottomLeft: { x, y: bottomBounds },
  }
}

const getBoundingCurvesFromRectangularBounds = ({ x, y, width, height }) => {
  const corners = getCornerPoints(x, y, width, height)

  const boundingCurves = {
    top: {
      startPoint: corners.topLeft,
      controlPoint1: corners.topLeft,
      controlPoint2: corners.topRight,
      endPoint: corners.topRight,
    },
    right: {
      startPoint: corners.topRight,
      controlPoint1: corners.topRight,
      controlPoint2: corners.bottomRight,
      endPoint: corners.bottomRight,
    },
    bottom: {
      startPoint: corners.bottomLeft,
      controlPoint1: corners.bottomLeft,
      controlPoint2: corners.bottomRight,
      endPoint: corners.bottomRight,
    },
    left: {
      startPoint: corners.topLeft,
      controlPoint1: corners.topLeft,
      controlPoint2: corners.bottomLeft,
      endPoint: corners.bottomLeft,
    },
  }

  return boundingCurves
}

// -----------------------------------------------------------------------------
// Exports
// -----------------------------------------------------------------------------

export const getRandomBoundingCurves = (canvas) => {
  const bounds = getRandomRectangleBounds(
    canvas.width,
    canvas.height,
    SHAPE_MIN_DISTANCE_FROM_EDGE
  )

  const boundingCurves = getBoundingCurvesFromRectangularBounds(bounds)

  // Loop through each corner and expand the control points
  return reduce(
    (acc, nodeId) => expandBoundingCurvesCornerControlPoints(nodeId, acc),
    boundingCurves,
    CORNER_POINTS
  )
}

export const clampGridSquareToGridDimensions = curry(
  ({ columns, rows }, { x, y }) => {
    const columnsTotal = isInteger(columns) ? columns : columns.length
    const rowsTotal = isInteger(rows) ? rows : rows.length

    const gridSquare = {}

    const xInt = parseInt(x)
    const yInt = parseInt(y)

    if (isInteger(xInt)) {
      gridSquare.x = clampNumberBetween(0, columnsTotal - 1, x)
    }

    if (isInteger(yInt)) {
      gridSquare.y = clampNumberBetween(0, rowsTotal - 1, y)
    }

    return gridSquare
  }
)
