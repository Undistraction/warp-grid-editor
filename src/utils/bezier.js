import { Bezier } from 'bezier-js'
import { pipe, reduce } from 'ramda'

// -----------------------------------------------------------------------------
// Utils
// -----------------------------------------------------------------------------

const getBezier = ({ startPoint, endPoint, controlPoint1, controlPoint2 }) => {
  const args = [
    startPoint.x,
    startPoint.y,
    controlPoint1.x,
    controlPoint1.y,
    controlPoint2.x,
    controlPoint2.y,
    endPoint.x,
    endPoint.y,
  ]

  return new Bezier(...args)
}

// -----------------------------------------------------------------------------
// Exports
// -----------------------------------------------------------------------------

export const getBounds = pipe(
  Object.values,
  reduce(
    (acc, curve) => {
      const bounds = getBezier(curve).bbox()
      return {
        xMax: Math.max(acc.xMax, bounds.x.max),
        yMax: Math.max(acc.yMax, bounds.y.max),
        xMin: Math.min(acc.xMin, bounds.x.min),
        yMin: Math.min(acc.yMin, bounds.y.min),
      }
    },
    { xMax: 0, yMax: 0, xMin: Infinity, yMin: Infinity }
  )
)
