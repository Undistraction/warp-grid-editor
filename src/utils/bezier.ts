import { Bezier } from 'bezier-js'
import { pipe, reduce } from 'ramda'

import type { Curve } from '../types'

// -----------------------------------------------------------------------------
// Types
// -----------------------------------------------------------------------------

type Bounds = {
  xMax: number
  yMax: number
  xMin: number
  yMin: number
}

type BezierArgs = [
  number,
  number,
  number,
  number,
  number,
  number,
  number,
  number,
]

// -----------------------------------------------------------------------------
// Utils
// -----------------------------------------------------------------------------

const getBezier = ({
  startPoint,
  endPoint,
  controlPoint1,
  controlPoint2,
}: Curve) => {
  const args: BezierArgs = [
    startPoint.x,
    startPoint.y,
    controlPoint1.x,
    controlPoint1.y,
    controlPoint2.x,
    controlPoint2.y,
    endPoint.x,
    endPoint.y,
  ]

  return new Bezier(...args)
}

// -----------------------------------------------------------------------------
// Exports
// -----------------------------------------------------------------------------

export const getBounds = pipe(
  Object.values,
  reduce(
    (acc: Bounds, curve: Curve) => {
      const bounds = getBezier(curve).bbox()
      return {
        xMax: Math.max(acc.xMax, bounds.x.max),
        yMax: Math.max(acc.yMax, bounds.y.max),
        xMin: Math.min(acc.xMin, bounds.x.min),
        yMin: Math.min(acc.yMin, bounds.y.min),
      }
    },
    { xMax: 0, yMax: 0, xMin: Infinity, yMin: Infinity }
  )
)
