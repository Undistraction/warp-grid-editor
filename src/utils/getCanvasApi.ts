import { isArray } from 'ramda-adjunct'

import type {
  BoundingCurves,
  Curve,
  Point,
  StraightLine,
  WarpGrid,
} from '../types'

// -----------------------------------------------------------------------------
// Utils
// -----------------------------------------------------------------------------

const isCurve = (value: Curve | StraightLine) => {
  return `controlPoint1` in value || `controlPoint2` in value
}

// -----------------------------------------------------------------------------
// Exports
// -----------------------------------------------------------------------------

const getCanvasApi = (context: CanvasRenderingContext2D) => {
  const drawDot = ({ x, y }: Point, { color = `black`, size = 2 } = {}) => {
    context.beginPath()
    context.moveTo(x, y)
    context.arc(x, y, size, 0, Math.PI * 2)
    context.fillStyle = color
    context.fill()
  }

  const drawCurve = (
    { controlPoint1, controlPoint2, startPoint, endPoint }: Curve,
    { lineColor = `black`, lineWidth = 1 } = {}
  ) => {
    context.beginPath()
    context.moveTo(startPoint.x, startPoint.y)

    // Draw curve from point 0 to point 1
    context.bezierCurveTo(
      controlPoint1.x,
      controlPoint1.y,
      controlPoint2.x,
      controlPoint2.y,
      endPoint.x,
      endPoint.y
    )

    context.strokeStyle = lineColor
    context.lineWidth = lineWidth
    context.stroke()
  }

  const drawQuad = (
    boundingCurves: BoundingCurves,
    { color = `black`, lineWidth = 1, fill = `transparent` } = {}
  ) => {
    context.beginPath()
    context.moveTo(
      boundingCurves.top.startPoint.x,
      boundingCurves.top.startPoint.y
    )

    // Draw Top (left to right)
    context.bezierCurveTo(
      boundingCurves.top.controlPoint1.x,
      boundingCurves.top.controlPoint1.y,
      boundingCurves.top.controlPoint2.x,
      boundingCurves.top.controlPoint2.y,
      boundingCurves.top.endPoint.x,
      boundingCurves.top.endPoint.y
    )

    // Draw Right (top to bottom)
    context.bezierCurveTo(
      boundingCurves.right.controlPoint1.x,
      boundingCurves.right.controlPoint1.y,
      boundingCurves.right.controlPoint2.x,
      boundingCurves.right.controlPoint2.y,
      boundingCurves.right.endPoint.x,
      boundingCurves.right.endPoint.y
    )

    // Draw Bottom (right to left) note the changed order as the line is
    // defined left to right and we are drawing right to left
    context.bezierCurveTo(
      boundingCurves.bottom.controlPoint2.x,
      boundingCurves.bottom.controlPoint2.y,
      boundingCurves.bottom.controlPoint1.x,
      boundingCurves.bottom.controlPoint1.y,
      boundingCurves.bottom.startPoint.x,
      boundingCurves.bottom.startPoint.y
    )

    // Draw Left (bottom to top) note the changed order as the line is
    // defined top to bottom and we are drawing bottom to top
    context.bezierCurveTo(
      boundingCurves.left.controlPoint2.x,
      boundingCurves.left.controlPoint2.y,
      boundingCurves.left.controlPoint1.x,
      boundingCurves.left.controlPoint1.y,
      boundingCurves.left.startPoint.x,
      boundingCurves.left.startPoint.y
    )

    context.closePath()
    context.strokeStyle = color
    context.fillStyle = fill
    context.lineWidth = lineWidth
    context.fill()
    context.stroke()
  }

  const drawPatchBounds = (
    boundingCurves: BoundingCurves,
    { lineColor = `black`, lineWidth = 3, fill = `transparent` } = {}
  ) => {
    drawQuad(boundingCurves, { color: lineColor, lineWidth, fill })
  }

  const drawBounds = (
    boundingCurves: BoundingCurves,
    { lineColor = `black`, lineWidth = 2 } = {}
  ) => {
    drawQuad(boundingCurves, { color: lineColor, lineWidth })
  }

  const clearCanvas = (canvas: HTMLCanvasElement) => {
    context.clearRect(0, 0, canvas.width, canvas.height)
  }

  const drawGridCurve = (
    line: Curve | StraightLine,
    { lineColor = `black` } = {}
  ) => {
    if (isCurve(line)) {
      drawCurve(line, { lineColor })
    } else {
      drawStraightLine(line, { lineColor })
    }
  }

  const drawCoonsPatch = (
    grid: WarpGrid,
    {
      shouldDrawIntersections,
      shouldDrawBounds,
      shouldDrawGrid,
    }: {
      shouldDrawIntersections: boolean
      shouldDrawBounds: boolean
      shouldDrawGrid: boolean
    }
  ) => {
    if (shouldDrawBounds) {
      drawBounds(grid.model.boundingCurves)
    }

    if (shouldDrawGrid) {
      const lines = grid.getLines()
      // Draw lines along x axis
      lines.xAxis.map((curveSectionsOrCurve) => {
        if (isArray(curveSectionsOrCurve)) {
          curveSectionsOrCurve.map((curve) =>
            drawGridCurve(curve, { lineColor: `rgba(0,0,0,0.7)` })
          )
        } else {
          drawGridCurve(curveSectionsOrCurve, { lineColor: `rgba(0,0,0,0.7)` })
        }
      })

      //Draw lines along y axis
      lines.yAxis.map((curveSectionsOrCurve) => {
        if (isArray(curveSectionsOrCurve)) {
          curveSectionsOrCurve.map((curve) =>
            drawGridCurve(curve, { lineColor: `rgba(0,0,0,0.7)` })
          )
        } else {
          drawGridCurve(curveSectionsOrCurve, { lineColor: `rgba(0,0,0,0.7)` })
        }
      })
    }

    if (shouldDrawIntersections) {
      // Draw intersections between grid boundingCurves
      grid.getIntersections().map((point) => {
        drawDot(point, {
          size: 3,
        })
      })
    }
  }

  const drawGridSquareBounds = (boundingCurves: BoundingCurves) => {
    drawPatchBounds(boundingCurves, {
      fill: `black`,
      lineWidth: 0,
    })
  }

  const drawStraightLine = (
    line: StraightLine,
    { lineColor = `black`, lineWidth = 1 } = {}
  ) => {
    context.beginPath()
    context.moveTo(line.startPoint.x, line.startPoint.y)
    context.lineTo(line.endPoint.x, line.endPoint.y)
    context.strokeStyle = lineColor
    context.lineWidth = lineWidth
    context.stroke()
  }

  return {
    clearCanvas,
    drawCoonsPatch,
    drawGridSquareBounds,
  }
}

export default getCanvasApi
