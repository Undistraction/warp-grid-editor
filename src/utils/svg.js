import { flatten, map } from 'ramda'

// -----------------------------------------------------------------------------
// Utils
// -----------------------------------------------------------------------------

const getMoveForPoint = ({ x, y }) => `M${x} ${y}`

const getCubicBezierToPoint = ({ controlPoint1, controlPoint2, endPoint }) =>
  `C${controlPoint1.x} ${controlPoint1.y} ${controlPoint2.x} ${controlPoint2.y} ${endPoint.x} ${endPoint.y}`

const getPathForBounds = ({ top, bottom, left, right }) => {
  const bottomReversed = {
    endPoint: bottom.startPoint,
    controlPoint1: bottom.controlPoint2,
    controlPoint2: bottom.controlPoint1,
  }

  const leftReversed = {
    endPoint: left.startPoint,
    controlPoint1: left.controlPoint2,
    controlPoint2: left.controlPoint1,
  }

  return `${getMoveForPoint(top.startPoint)} ${getCubicBezierToPoint(top)} ${getCubicBezierToPoint(right)} ${getCubicBezierToPoint(bottomReversed)} ${getCubicBezierToPoint(leftReversed)} Z`
}

// -----------------------------------------------------------------------------
// Exports
// -----------------------------------------------------------------------------

// Run through each subcell, rendering the palette colours
export const getSvgForGrid = (grid, { width, height }) => {
  const gridCells = grid.getAllCellBounds()
  const gridCellsFlat = flatten(gridCells)

  const elements = map((cell) => {
    const paths = getPathForBounds(cell)
    return `<path stroke="black" fill="transparent" d="${paths}" />`
  }, gridCellsFlat)

  return `<svg
  xmlns="http://www.w3.org/2000/svg"
  overflow="visible"
  viewBox="0 0 ${width} ${height}"
  width="${width}px"
  height="${height}px"
>
  ${elements.join(`\n  `)}
</svg>`
}
