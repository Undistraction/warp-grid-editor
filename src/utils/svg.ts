import { flatten, map } from 'ramda'

import type { BoundingCurves, Curve, Point, Size, WarpGrid } from '../types'

// -----------------------------------------------------------------------------
// Exports
// -----------------------------------------------------------------------------

export const getMoveForPoint = ({ x, y }: Point) => `M ${x} ${y}`

export const getCubicBezierToPoint = ({
  controlPoint1,
  controlPoint2,
  endPoint,
}: Curve) => {
  return `C ${controlPoint1.x} ${controlPoint1.y} ${controlPoint2.x} ${controlPoint2.y} ${endPoint.x} ${endPoint.y}`
}

export const getPathForBounds = ({
  top,
  bottom,
  left,
  right,
}: BoundingCurves) => {
  const bottomReversed = {
    startPoint: bottom.endPoint,
    endPoint: bottom.startPoint,
    controlPoint1: bottom.controlPoint2,
    controlPoint2: bottom.controlPoint1,
  }

  const leftReversed = {
    startPoint: left.endPoint,
    endPoint: left.startPoint,
    controlPoint1: left.controlPoint2,
    controlPoint2: left.controlPoint1,
  }

  return `${getMoveForPoint(top.startPoint)} ${getCubicBezierToPoint(top)} ${getCubicBezierToPoint(right)} ${getCubicBezierToPoint(bottomReversed)} ${getCubicBezierToPoint(leftReversed)} Z`
}

// Run through each subcell, rendering the palette colours
export const getSvgForGrid = (grid: WarpGrid, { width, height }: Size) => {
  const gridCells = grid.getAllCellBounds()
  const gridCellsFlat = flatten(gridCells)

  const elements = map((cell) => {
    const paths = getPathForBounds(cell)
    return `<path stroke="black" fill="transparent" d="${paths}" />`
  }, gridCellsFlat)

  return `<svg
  xmlns="http://www.w3.org/2000/svg"
  overflow="visible"
  viewBox="0 0 ${width} ${height}"
  width="${width}px"
  height="${height}px"
>
  ${elements.join(`\n  `)}
</svg>`
}
