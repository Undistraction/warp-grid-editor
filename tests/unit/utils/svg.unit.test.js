import warpGrid from 'warp-grid'

import { getSvgForGrid } from '../../../src/utils/svg'

// -----------------------------------------------------------------------------
// Const
// -----------------------------------------------------------------------------

const BOUNDING_CURVES = {
  top: {
    startPoint: { x: 0, y: 0 },
    endPoint: { x: 100, y: 0 },
    controlPoint1: { x: 10, y: -10 },
    controlPoint2: { x: 90, y: -10 },
  },
  bottom: {
    startPoint: { x: 0, y: 100 },
    endPoint: { x: 100, y: 100 },
    controlPoint1: { x: -10, y: 110 },
    controlPoint2: { x: 110, y: 110 },
  },
  left: {
    startPoint: { x: 0, y: 0 },
    endPoint: { x: 0, y: 100 },
    controlPoint1: { x: -10, y: -10 },
    controlPoint2: { x: -10, y: 110 },
  },
  right: {
    startPoint: { x: 100, y: 0 },
    endPoint: { x: 100, y: 100 },
    controlPoint1: { x: 110, y: -10 },
    controlPoint2: { x: 110, y: 110 },
  },
}

// -----------------------------------------------------------------------------
// Tests
// -----------------------------------------------------------------------------

describe(`getSvgForGrid`, () => {
  it(`returns an svg document for the supplied grid`, () => {
    const grid = warpGrid(BOUNDING_CURVES, { columns: 3, rows: 5 })
    expect(getSvgForGrid(grid, { width: 110, height: 110 })).toMatchSnapshot()
  })
})
