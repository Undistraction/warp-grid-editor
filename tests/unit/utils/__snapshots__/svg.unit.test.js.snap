// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getSvgForGrid > returns an svg document for the supplied grid 1`] = `
"<svg
  xmlns="http://www.w3.org/2000/svg"
  overflow="visible"
  viewBox="0 0 110 110"
  width="110px"
  height="110px"
>
  <path stroke="black" fill="transparent" d="M 0 0 C 0 0 32.93531070818844 -7.003774504353478 32.93531070818844 -7.003774504353478 C 32.93531070818844 -7.003774504353478 30.731112487204886 13.945938005902146 30.731112487204886 13.945938005902146 C 30.731112487204886 13.945938005902146 -6.211679114533506 18.120956058294 -6.211679114533506 18.120956058294 C -6.211679114533506 18.120956058294 0 0 0 0 Z" />
  <path stroke="black" fill="transparent" d="M 32.93531070818844 -7.003774504353478 C 32.93531070818844 -7.003774504353478 67.06468929181153 -7.003774504353482 67.06468929181153 -7.003774504353482 C 67.06468929181153 -7.003774504353482 69.26888751279508 13.945938005902148 69.26888751279508 13.945938005902148 C 69.26888751279508 13.945938005902148 30.731112487204886 13.945938005902146 30.731112487204886 13.945938005902146 C 30.731112487204886 13.945938005902146 32.93531070818844 -7.003774504353478 32.93531070818844 -7.003774504353478 Z" />
  <path stroke="black" fill="transparent" d="M 67.06468929181153 -7.003774504353482 C 67.06468929181153 -7.003774504353482 100 0 100 0 C 100 0 106.21167911453352 18.120956058294 106.21167911453352 18.120956058294 C 106.21167911453352 18.120956058294 69.26888751279508 13.945938005902148 69.26888751279508 13.945938005902148 C 69.26888751279508 13.945938005902148 67.06468929181153 -7.003774504353482 67.06468929181153 -7.003774504353482 Z" />
  <path stroke="black" fill="transparent" d="M -6.211679114533506 18.120956058294 C -6.211679114533506 18.120956058294 30.731112487204886 13.945938005902146 30.731112487204886 13.945938005902146 C 30.731112487204886 13.945938005902146 30.210248756531833 38.00448538835119 30.210248756531833 38.00448538835119 C 30.210248756531833 38.00448538835119 -7.373354758135497 39.350746988781424 -7.373354758135497 39.350746988781424 C -7.373354758135497 39.350746988781424 -6.211679114533506 18.120956058294 -6.211679114533506 18.120956058294 Z" />
  <path stroke="black" fill="transparent" d="M 30.731112487204886 13.945938005902146 C 30.731112487204886 13.945938005902146 69.26888751279508 13.945938005902148 69.26888751279508 13.945938005902148 C 69.26888751279508 13.945938005902148 69.78975124346813 38.00448538835117 69.78975124346813 38.00448538835117 C 69.78975124346813 38.00448538835117 30.210248756531833 38.00448538835119 30.210248756531833 38.00448538835119 C 30.210248756531833 38.00448538835119 30.731112487204886 13.945938005902146 30.731112487204886 13.945938005902146 Z" />
  <path stroke="black" fill="transparent" d="M 69.26888751279508 13.945938005902148 C 69.26888751279508 13.945938005902148 106.21167911453352 18.120956058294 106.21167911453352 18.120956058294 C 106.21167911453352 18.120956058294 107.37335475813549 39.350746988781424 107.37335475813549 39.350746988781424 C 107.37335475813549 39.350746988781424 69.78975124346813 38.00448538835117 69.78975124346813 38.00448538835117 C 69.78975124346813 38.00448538835117 69.26888751279508 13.945938005902148 69.26888751279508 13.945938005902148 Z" />
  <path stroke="black" fill="transparent" d="M -7.373354758135497 39.350746988781424 C -7.373354758135497 39.350746988781424 30.210248756531833 38.00448538835119 30.210248756531833 38.00448538835119 C 30.210248756531833 38.00448538835119 30.07661024039278 62.13174786275002 30.07661024039278 62.13174786275002 C 30.07661024039278 62.13174786275002 -7.373354758135496 60.6492530112186 -7.373354758135496 60.6492530112186 C -7.373354758135496 60.6492530112186 -7.373354758135497 39.350746988781424 -7.373354758135497 39.350746988781424 Z" />
  <path stroke="black" fill="transparent" d="M 30.210248756531833 38.00448538835119 C 30.210248756531833 38.00448538835119 69.78975124346813 38.00448538835117 69.78975124346813 38.00448538835117 C 69.78975124346813 38.00448538835117 69.9233897596072 62.131747862750004 69.9233897596072 62.131747862750004 C 69.9233897596072 62.131747862750004 30.07661024039278 62.13174786275002 30.07661024039278 62.13174786275002 C 30.07661024039278 62.13174786275002 30.210248756531833 38.00448538835119 30.210248756531833 38.00448538835119 Z" />
  <path stroke="black" fill="transparent" d="M 69.78975124346813 38.00448538835117 C 69.78975124346813 38.00448538835117 107.37335475813549 39.350746988781424 107.37335475813549 39.350746988781424 C 107.37335475813549 39.350746988781424 107.37335475813549 60.6492530112186 107.37335475813549 60.6492530112186 C 107.37335475813549 60.6492530112186 69.9233897596072 62.131747862750004 69.9233897596072 62.131747862750004 C 69.9233897596072 62.131747862750004 69.78975124346813 38.00448538835117 69.78975124346813 38.00448538835117 Z" />
  <path stroke="black" fill="transparent" d="M -7.373354758135496 60.6492530112186 C -7.373354758135496 60.6492530112186 30.07661024039278 62.13174786275002 30.07661024039278 62.13174786275002 C 30.07661024039278 62.13174786275002 30.330196938787726 86.19029524519904 30.330196938787726 86.19029524519904 C 30.330196938787726 86.19029524519904 -6.211679114533504 81.87904394170602 -6.211679114533504 81.87904394170602 C -6.211679114533504 81.87904394170602 -7.373354758135496 60.6492530112186 -7.373354758135496 60.6492530112186 Z" />
  <path stroke="black" fill="transparent" d="M 30.07661024039278 62.13174786275002 C 30.07661024039278 62.13174786275002 69.9233897596072 62.131747862750004 69.9233897596072 62.131747862750004 C 69.9233897596072 62.131747862750004 69.66980306121226 86.190295245199 69.66980306121226 86.190295245199 C 69.66980306121226 86.190295245199 30.330196938787726 86.19029524519904 30.330196938787726 86.19029524519904 C 30.330196938787726 86.19029524519904 30.07661024039278 62.13174786275002 30.07661024039278 62.13174786275002 Z" />
  <path stroke="black" fill="transparent" d="M 69.9233897596072 62.131747862750004 C 69.9233897596072 62.131747862750004 107.37335475813549 60.6492530112186 107.37335475813549 60.6492530112186 C 107.37335475813549 60.6492530112186 106.21167911453347 81.87904394170602 106.21167911453347 81.87904394170602 C 106.21167911453347 81.87904394170602 69.66980306121226 86.190295245199 69.66980306121226 86.190295245199 C 69.66980306121226 86.190295245199 69.9233897596072 62.131747862750004 69.9233897596072 62.131747862750004 Z" />
  <path stroke="black" fill="transparent" d="M -6.211679114533504 81.87904394170602 C -6.211679114533504 81.87904394170602 30.330196938787726 86.19029524519904 30.330196938787726 86.19029524519904 C 30.330196938787726 86.19029524519904 32.26711812749319 107.14000775545462 32.26711812749319 107.14000775545462 C 32.26711812749319 107.14000775545462 0 100 0 100 C 0 100 -6.211679114533504 81.87904394170602 -6.211679114533504 81.87904394170602 Z" />
  <path stroke="black" fill="transparent" d="M 30.330196938787726 86.19029524519904 C 30.330196938787726 86.19029524519904 69.66980306121226 86.190295245199 69.66980306121226 86.190295245199 C 69.66980306121226 86.190295245199 67.73288187250682 107.14000775545463 67.73288187250682 107.14000775545463 C 67.73288187250682 107.14000775545463 32.26711812749319 107.14000775545462 32.26711812749319 107.14000775545462 C 32.26711812749319 107.14000775545462 30.330196938787726 86.19029524519904 30.330196938787726 86.19029524519904 Z" />
  <path stroke="black" fill="transparent" d="M 69.66980306121226 86.190295245199 C 69.66980306121226 86.190295245199 106.21167911453347 81.87904394170602 106.21167911453347 81.87904394170602 C 106.21167911453347 81.87904394170602 100 100 100 100 C 100 100 67.73288187250682 107.14000775545463 67.73288187250682 107.14000775545463 C 67.73288187250682 107.14000775545463 69.66980306121226 86.190295245199 69.66980306121226 86.190295245199 Z" />
</svg>"
`;
